### PR TITLE
feat: extend oci.Store with DeleteTree().

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi!

Following the merge of #614, I opened this PR to propose `DeleteTree()`.
This function will delete the content matching the given descriptor but also its children if they do not have any other parent.
We use this code in Inspektor Gadget to delete OCI image:
https://github.com/inspektor-gadget/inspektor-gadget/pull/2162
https://github.com/inspektor-gadget/inspektor-gadget/pull/2162/commits/248811856471fb36f4befe36af77a27ee57249b2#diff-570aea5baf5f0e0b7f46b570a48f173e8858df3436f091fb61d2e3c745db6f0eR298
While it is still not merged at the moment, it was nonetheless validated.

If you see any way to improve this PR, feel free to share your feedback.

Best regards and thank you in advance.

### CODE REVIEW

File: content/oci/oci.go [hunk 1]
```diff
@@ -168,6 +168,34 @@ func (s *Store) Delete(ctx context.Context, target ocispec.Descriptor) error {
...
+		return fmt.Errorf("getting image predecessors: %w", err)
+	}
+
+	if len(predecessors) > 1 {
+		return nil
+	}
+
+	descriptors, err := content.Successors(ctx, s, root)
+	if err != nil {
+		return fmt.Errorf("getting image successors: %w", err)
```
Comment (TerryHowe) [inline content/oci/oci.go:189]:
```suggestion
		return fmt.Errorf("getting successors: %w", err)
```

File: content/oci/oci.go [hunk 2]
```diff
@@ -168,6 +168,34 @@ func (s *Store) Delete(ctx context.Context, target ocispec.Descriptor) error {
...
+// DeleteTree deletes the content matching the descriptor as well as its
+// children from the store.
+// Children will be deleted only if they do not any other predecessors than the
+// given root.
+// DeleteTree may fail on certain systems (i.e. NTFS), if there is a process
+// (i.e. an unclosed Reader) using any of root children.
+func (s *Store) DeleteTree(ctx context.Context, root ocispec.Descriptor) error {
+	predecessors, err := s.Predecessors(ctx, root)
+	if err != nil {
+		return fmt.Errorf("getting image predecessors: %w", err)
```
Comment (TerryHowe) [inline content/oci/oci.go:180]:
```suggestion
		return fmt.Errorf("getting predecessors: %w", err)
```
Or maybe something other than "image" I would think

**Comment by shizhMSFT** [comment]
@eiffel-fl Thank you for proposing the `DeleteTree()`.

Instead of deleting a tree, I'm thinking about garbage collection (GC).

There are many benefits of GC over simply deleting a tree.
- GC can be performed after several combination of `Push()` and `Delete()` operations.
- GC can clean up any dangling nodes.

The basic idea is to mimic the garbage collection behavior of a real registry.

Related:
- #472

/cc @wangxiaoxuan273 @Wwwsylvia

**Comment by wangxiaoxuan273** [comment]
@eiffel-fl Thanks for opening this PR. Let's make future discussions on the #472 issue page, so that more stakeholders can participate.

**Comment by Wwwsylvia** [comment]
@eiffel-fl Given that #653 has been merged, shall we close this PR? Thanks for bringing this up!

**Comment by Wwwsylvia** [comment]
Hi @eiffel-fl , can you help us to understand the relationship between those blobs you listed?

```sh
$sudo ls /var/lib/ig/oci-store/blobs/sha256  
225a8e697e260684d9451ccf8a622cd5548b0ae5fac6e13c416cd1d85247b7b7  b4be31f07444fb9be6dee8cd1349aab3c780d333b63406be10c5c14c1b39dd37
662ca58536d5a03864781eb42870f6712927981383f1dab1449d9adf80067313  c06a86dad7e582686b53bec9c729b92fc56c838b59cc11d0105c2c20f3158dcd
b0eff7f0ec2a27b8d71fe904cf373fa2002c6d669db54c25a9542cf2cb5f46d1  f61eef5ff800b2e4366c6d2e7311e95ce8dc3cff5c020a3a810429656fd93efb
```

A [mermaid diagram](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) would be very helpful.

**Comment by Wwwsylvia** [comment]
@eiffel-fl I have some further questions:

- What are their media types?
- Is any of them a `subject` of another one?
- Which of them are recorded in `index.json`?

**Comment by Wwwsylvia** [comment]
Hi @eiffel-fl , thanks for the clarification. I now understand what's happening.

**TL;DR: the behavior you observed is by design. We don't automatically delete the successor manifests along with the deleted index, if they are present in `index.json`.**

Let me put the diagram in another way:

```mermaid
graph TD;
b0eff7f0ec2a27b8d71fe904cf373fa2002c6d669db54c25a9542cf2cb5f46d1[Index:latest] --> 662ca58536d5a03864781eb42870f6712927981383f1dab1449d9adf80067313
b0eff7f0ec2a27b8d71fe904cf373fa2002c6d669db54c25a9542cf2cb5f46d1 --> f61eef5ff800b2e4366c6d2e7311e95ce8dc3cff5c020a3a810429656fd93efb
662ca58536d5a03864781eb42870f6712927981383f1dab1449d9adf80067313[Manifest:amd64] --> b4be31f07444fb9be6dee8cd1349aab3c780d333b63406be10c5c14c1b39dd37
662ca58536d5a03864781eb42870f6712927981383f1dab1449d9adf80067313 --> 225a8e697e260684d9451ccf8a622cd5548b0ae5fac6e13c416cd1d85247b7b7
f61eef5ff800b2e4366c6d2e7311e95ce8dc3cff5c020a3a810429656fd93efb[Manifest:arm64] --> b4be31f07444fb9be6dee8cd1349aab3c780d333b63406be10c5c14c1b39dd37
f61eef5ff800b2e4366c6d2e7311e95ce8dc3cff5c020a3a810429656fd93efb --> c06a86dad7e582686b53bec9c729b92fc56c838b59cc11d0105c2c20f3158dcd 

```

We know that:
-  `Index:latest` references two manifests: `Manifest:amd64` and `Manifest:arm64`
- `Index:latest`, `Manifest:amd64`, and `Manifest:arm64` are all recorded in `index.json`

When one is deleting `Index:latest`, they may expect:
- Manifest:amd64` and `Manifest:arm64` also get deleted
- Manifest:amd64` and `Manifest:arm64` are not deleted (they wants to manually delete `Manifest:amd64` later and keep `Manifest:arm64`)

`oras-go` could not tell the user's purpose, so it chooses not to automatically delete `Manifest:amd64`  and `Manifest:arm64`.
To delete the whole graph, you need to manually delete `Manifest:amd64` and `Manifest:arm64` after deleting `Index:latest`.

The idea is that `oras-go` considers everything in `index.json` is meaningful.
That is to say, if `Manifest:amd64` and `Manifest:arm64` were not present in `index.json`, they would get automatically deleted together with `Index:latest`.

Does this make sense to you?

**Comment by Wwwsylvia** [comment]
BTW, may I know what tool did you use to generate this OCI layout?

**Comment by Wwwsylvia** [comment]
> > 2. Is any of them a `subject` of another one?
> 
> Can you please precise how can I get this information 😅?

Oh you can just look into the manifests and see if they have a `subject` field, like:
```sh
cat /var/lib/ig/oci-store/blobs/sha256/662ca58536d5a03864781eb42870f6712927981383f1dab1449d9adf80067313 | jq .
```

But in this particular case, the answer is that: none of them is a [`subject`](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/manifest.md#image-manifest-property-descriptions) of another. 😀

**Comment by Wwwsylvia** [comment]
> Now, I am really wondering if we do not do something wrong as I find it odd to have the architecture manifests in index.json, as, from my understanding, the index.json should only contain index which do the abstraction for the architectures.

Actually, it is `oras-go` to blame!
Currently `oras-go` will automatically add manifests into `index.json`, no matter it has a tag or not. This was to make sure referrers manifests without tags were still trackable.
We might need to improve this design.

**Comment by Wwwsylvia** [comment]
This is not a bug, and it is not straightforward to change it🤔.
We need to work out a more proper design for this. Would you like to open an issue for discussion?

**Comment by Wwwsylvia** [comment]
@eiffel-fl Thank you!
